### PR TITLE
:sparkles: Add manual expiry to Redis keys

### LINKED ISSUE #93: Fix redis expiries
We have the following problems:

- `treasuries` is often evicted because of memory limits and we are just not making use of it.
- `whitelist` is often evicted and thus whitelists are probably fetch too many times for nothing.
- The `covalent_transfer_items`, `covalent_treasury`, `bitquery_eth` and `covalent_prices` hashmaps are getting expired often, because of memory limits. Whenever one of these hashes is picked by the eviction algorithm, the entirety of the hash is evicted (and not the oldest keys).

Let's break things down by kind of data:

### Treasuries

Expectations:

- Append new treasuries to the set whenever a new request is done.
- Regularly evict old treasuries.

Solution:

- Host the `treasuries` key in a Redis instance that doesn't have a broad eviction rule.
- Whenever a new Treasury is added, create a key with the address of the treasury (like `treasury_{addr}`, with empty string value and an expiry of, eg, 14 days.
- Whenever a Treasury is accessed (but already in Redis), update the expiry to now + 14 days.

### Whitelist

Expectations:

- Update the Redis `whitelist` object regularly, once a day. But not more often, as this would be useless use of bandwidth and create I/O delays.
- We don't want this to be append-only, but to be _replaced entirely_ on update.

Solution:

- Host the `whitelist` key in a Redis instance that doesn't have a broad eviction rule (same as `treasuries`).
- Define expiry to 1d.
- Replace the whole key contents on each update (do not use `sadd`, we could also just use a simple raw text key).

### Data

For all other pieces of data.

Expectations:

- All pieces of data should have an expiry of 1 day at creation (because our data is stale after 1 day anyways).
- Set a broad eviction rule so as to make sure the instance stays up whenever memory runs low.
- When the eviction algorithm is run, it should evict only some pieces of data, not all of them at once.

Solution:

- Use a Redis instance with allkeys-lru.
- Split all the `covalent_transfer_items`, `covalent_treasury`, `bitquery_eth` and `covalent_prices` hashes in single entries, like `covalent_transfer_items_{treasury_address}_{contract_address}_{chain_id}_{date}`, etc.
- Define an expiry of 1 day at creation.

### PR DESCRIPTION
**Pull request by acemasterjb**
## Summary
Fixes #93

## Details
Expiries were explicitly set for data keys, each set to 1 day.

### CODE REVIEW

File: backend/app/treasury/adapters/covalent/portfolio_v2.py [hunk 1]
```diff
@@ -62,7 +61,14 @@ async def get_treasury(
             )
             raise
 
-        db.hset(CACHE_HASH_TREASURY, cache_key, json.dumps(portfolio_data))
+        db.set(cache_key, json.dumps(portfolio_data))
+        if db.ttl(cache_key) <= 0:
+            time_to_evict = datetime.now(tz=dateutil.tz.UTC) + timedelta(days=1)
+            time_to_evict_ts = time_to_evict.replace(
+                hour=0, minute=0, second=0, microsecond=0
+            ).timestamp()
```
Comment (lajarre) [inline backend/app/treasury/adapters/covalent/portfolio_v2.py:69]: This logic, namely calculating the eviction time, is used by several pieces of code in the exact same way. That's a good opportunity for factoring it into a helper function, which can be tested.

File: backend/app/treasury/actions.py [hunk 2]
```diff
@@ -21,22 +21,39 @@
...
 ) -> BalancesAtTransfers:
     "Returns series of balances defined at times of assets transfers"
 
-    async def get_transfers(asset: ERC20) -> list[Transfer]:
-        return (
-            await bitquery.get_eth_transfers(treasury.address)
-            if asset.token_symbol == "ETH"
-            else await get_token_transfers(treasury.address, asset.token_address)
-        )
+    transfers_and_end_balance = await make_transfers_and_end_balance_for_treasury(
```
Comment (lajarre) [inline backend/app/treasury/actions.py:53]: If I read correctly, your changed the logic at the  `if transfers_for_asset` clause at line 39. That means that you are effectively making sure that the `BalancesAtTransfers` object doesn't contain any asset that has an empty transfers list.

What's the reason for this change? How is reducing the `BalancesAtTransfers` helping? Can't we go along well with some empty `Transfers` lists?

File: backend/app/treasury/actions.py [hunk 3]
```diff
@@ -139,8 +156,10 @@ def update_treasury_assets_risk_contributions(
 
 
 async def build_treasury_with_assets(
-    treasury_address: str, chain_id: int, start: str, end: str
+    request_params: tuple[tuple[str, int], str, str]
```
[the remark this answers is not in the document]
Comment (acemasterjb) [reply backend/app/treasury/actions.py:159]: Sure, initially I wanted to add more params to `build_treasury_with_assets` and I thought packing the params into one would make for a cleaner function signature.

I have since removed these extra params but I still maintained that using a single param is cleaner for this function.

File: backend/app/treasury/actions.py (hunk 2 above)
Comment (lajarre) [reply backend/app/treasury/actions.py:53]: I believe that's more of a job for `BalanceAtTransfers` itself: its constructor should be able to handle empty transfers (indeed, it seems totally reasonable that there is 0 transfers on a given period). That would make the code overall more robust. But I'm fine with this version for now.

**Review by lajarre — APPROVED** [review]
